Refactor report-parser and add tests.

briangreenery opened the pull request:
I'm not sure whether this is something that you're interested in, but this change refactors report parsing. It breaks parsing into many functions and removes duplication. It also fixes a few bugs:
- Timestamps in v3 reports were not processed correctly (needed to multiply by 1000).
- Fixed typo in v2 reports (`diagnostice` -> `diagnostics`)
- Used `getBatteryVoltage` everywhere to get consistent numbers across all report types.

I also added tests to get to 100% code and branch coverage, though the tests aren't exhaustive.

briangreenery committed:
Refactor report-parser and add tests.

amcgee commented:
Nice, I will take a look at this today!

briangreenery commented:
Also I noticed that report-processor will reject any report that doesn't have a uuid. So it seems like the only reports that can currently work are v3 reports, but I'm not really sure what the story is here.

amcgee commented:
Short answers:
- That change needs to propagate a bunch of places, it's assumed to be in seconds for UI calculations.  I've actually already made this change locally along with implementing V4 reports (see below).
- V3 reports are the only ones that matter, and they will actually be superseded by V4 reports very soon.  We don't need backwards-compatibility yet, but it's good to get factorization and tests in place so we support it eventually.  We could theoretically drop the code that support V<4 reports since it should never be used in the wild.
- I'll have to run through the instances where this is used, it's possible there was a difference in the way the hardware devices reported their battery voltages that necessitated different processing (or it could have just been poorly factored)

Awesome, I <3 test coverage!

briangreenery commented:
> That change needs to propagate a bunch of places, it's assumed to be in seconds for UI calculations.

For this part, the number is being used to construct a javascript Date object. So I'm not sure other places need to change. Like, it's not storing a number of seconds, it's storing a Date object.

amcgee commented:
Yes, but it's (in this version) being used to create a malformed date object.  When it's read from the database for a RESTAPI call it's converted to an integer and then multiplied by 1000 client-side.  See monitor_client.js line 231.  

I have a fix for this that propagates correctly so I will open a separate pull request.  I don't know if we want to expose JavaScript Date objects via the REST API but am open to suggestions.

briangreenery commented:
Ahhh, ok. I think seconds since 1970 is just fine for a rest-api. Another alternative is an ISO 8601 time string like `"2015-10-26T22:45:27Z"`.

briangreenery committed:
Incorporate support for v4 reports

briangreenery commented:
I incorporated the changes from pull #15, added tests for v4 report parsing, and fixed what appeared to be a bug. I think maybe the check for whether there's another entry should be:

```javascript
index + 9 <= reportBytes.length
```

instead of

```javascript
index + 9 < reportBytes.length
```

amcgee commented:
The update looks good, thanks for incorporating the changes!  Merging.
